python_modules/dagster/dagster/core/instance/ref.py

```diff
@@ -180,23 +297,50 @@ def from_dir(base_dir, config_filename=DAGSTER_CONFIG_YAML_FILENAME, overrides=N
             config_value, "local_artifact_storage", defaults["local_artifact_storage"]
         )
 
-        run_storage_data = configurable_class_data_or_default(
-            config_value, "run_storage", defaults["run_storage"]
-        )
-
-        event_storage_data = configurable_class_data_or_default(
-            config_value, "event_log_storage", defaults["event_log_storage"]
-        )
-
         compute_logs_data = configurable_class_data_or_default(
             config_value,
             "compute_logs",
             defaults["compute_logs"],
         )
 
-        schedule_storage_data = configurable_class_data_or_default(
-            config_value, "schedule_storage", defaults["schedule_storage"]
-        )
+        if config_value.get("storage"):
+            storage_data = configurable_storage_data(config_value["storage"])
+
+            if isinstance(storage_data, CompositeStorageRef):
+                storage = storage_data
+                run_storage_data = storage.run_storage_data
+                event_storage_data = storage.event_storage_data
+                schedule_storage_data = storage.schedule_storage_data
+            else:
+                # The storage class is here, we have to rehydrate the class here, so that we can
+                # extract the individual storages to stash on the serialized instance ref
+                storage = storage_data.rehydrate()
```

Comment:
this is a weird artifact in that we have to construct storage_data for the individual storage components in order to be backwards compatible.

The alternative here, is that we construct some stub implementation that takes the unified storage config and creates the individual components from that config, but that runs the risk of instantiating the same storage class multiple times.